File: gym_leaders_territory/gym_leaders_territory.go
package main

import (
	"fmt"
)

func remove(s []int, i int) []int {
	var index = -1
	for j, thing := range s {
		if thing == i {
			index = j
			break
		}
	}

	s[index] = s[len(s)-1]
	return s[:len(s)-1]
}

func analyzeGraph(numNodes, numEdges, target int) string {
	graph := make([][]int, numNodes)
	var from = -1
	var to = -1
	for i := 0; i < numEdges; i++ {
		_, err := fmt.Scanln(&from, &to)
		if err != nil {
			fmt.Println("Error occurred:", err)
			return "err"
		}
		from--
		to--
		graph[from] = append(graph[from], to)
		graph[to] = append(graph[to], from)

	}
	madeDifference := true
	for madeDifference {
		if len(graph[target]) < 2 {
			return "yes"
		}
		madeDifference = false
		for i, connections := range graph {
			if len(connections) == 1 {
				graph[connections[0]] = remove(graph[connections[0]], i)
				graph[i] = []int{}
				madeDifference = true
			}
		}
	}
	return "no"
}

func main() {
	var numNodes = -1
	var numEdges = -1
	var target = -1
	_, err := fmt.Scanln(&numNodes, &target, &numEdges)
	if err != nil {
		fmt.Println("Error occurred:", err)
		return
	}

	result := analyzeGraph(numNodes, numEdges, target-1)
	fmt.Println(result)
}

File: gym_leaders_territory/gym_leaders_territory.py
def dfs(graph, node, visited):
    visited[node] = True
    for neighbor in graph[node]:
        if not visited[neighbor]:
            visited = dfs(graph, neighbor, visited)
    return visited


def can_conquer(num_gym_leaders, rival_leader, alliances):
    graph = [[] for _ in range(num_gym_leaders + 1)]

    for alliance in alliances:
        leader1, leader2 = alliance
        graph[leader1].append(leader2)
        graph[leader2].append(leader1)

    visited = [False] * (num_gym_leaders + 1)
    dfs(graph, rival_leader, visited)

    has_changed = True
    while has_changed:
        if len(graph[rival_leader]) <= 1:
            return "YES"

        has_changed = False
        for i, connections in enumerate(graph):
            if len(connections) == 1:
                graph[connections[0]].remove(i)
                graph[i] = []
                has_changed = True

        visited = [False] * (num_gym_leaders + 1)
        for i, leader in enumerate(dfs(graph, rival_leader, visited)):
            if not leader:
                graph[leader] = []

    return "NO"


def main():
    n, rival, m = map(int, input().split())
    alliances = [tuple(map(int, input().split())) for _ in range(m)]

    result = can_conquer(n, rival, alliances)
    print(result)


if __name__ == "__main__":
    main()
